### DISCUSSION TITLE
버퍼풀을 어떻게 관리해야할까요?

### DISCUSSION THREAD

**Discussion by HaiSeong** [woowacourse-study-2024/java-database#1] [General]
버퍼풀을 글로벌로 관리할지 또는 페이지 파일 인스턴스 마다 관리할지 고민중입니다.

제가 생각하는 각각의 장단점은 다음과 같습니다.

## 글로벌로 관리
- 장점
  - 전체 버퍼풀 영역을 효율적으로 관리 가능
- 단점
  - 구현이 복잡함

## 페이지 파일 인스턴스 마다 관리
- 장점
  - 각 파일 특성에 맞는 알고리즘으로 버퍼풀을 관리할 수 있음
- 단점
  - 각 파일별로 버퍼풀이 생성되어 메모리 관점에서 비효율적

**Comment by Chocochip101** [comment]
질문과 저의 의견입니다.

1. 글로벌로 관리하는게 구현이 복잡하나요? 저한테는 각 파일마다 버퍼풀을 생성해줘야 한다는 관점에서 구현할게 많아지고, 유지보수가 어려울 것 같습니다.
2. 페이지 파일 인스턴스 마다 관리하면 메모리 관점에서 비효율적인가요? 메모리 사용량만 둘 다 같게 통제한다면, 비효율적인지는 모르겠습니다.

**Reply by HaiSeong** [reply]
1, 2번 답변을 보고 생각해보니 오히려 '페이지 파일 인스턴스 마다 관리' 하는 방식이 훨신 복잡해 보이네요.

각각의 파일마다 버퍼풀이 따로 관리되면 버퍼풀의 전체 크기를 관리 해줘야 한다는 점에서 구현하기 까다로울것 같아요.

추후에 특정 테이블에서 많은 페이지를 읽어온다 가정 했을때  특정 파일에 더 많은 버퍼를 할당할 수 있어 글로벌로 관리하는게 좋아보이네요.

그렇다면 버퍼풀에서 페이지를 구분할때 `(파일명, 페이지 번호)`로 구분해야 할까요?

**Reply by Chocochip101** [reply]
네, 말씀해 주신 대로 구분해야 할 것 같아요. 여기서 관건은 파일명을 어떻게 설정할지겠네요 :)

**Reply by HaiSeong** [reply]
초코칩은 신입니까?
